precompiles/utils/src/xcm.rs

```diff
@@ -0,0 +1,465 @@
+                encoded.push(length);
+                encoded.append(&mut data.into());
+                encoded.as_slice().into()
+            }
+            Junction::OnlyChild => {
+                encoded.push(7u8);
+                encoded.as_slice().into()
+            }
+            Junction::GlobalConsensus(network_id) => {
+                encoded.push(9u8);
+                encoded.append(&mut network_id_to_bytes(Some(network_id)));
+                encoded.as_slice().into()
+            }
+            // TODO: The only missing item here is Junciton::Plurality. This is a complex encoded
+            // type that we need to evaluate how to support
+            _ => unreachable!("Junction::Plurality not supported yet"),
+        };
+        EvmData::write(writer, encoded_bytes);
+    }
+
+    fn has_static_size() -> bool {
+        false
+    }
+}
+
+impl EvmData for Junctions {
+    fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+        let junctions_bytes: Vec<Junction> = reader.read()?;
+        let mut junctions = Junctions::Here;
+        for item in junctions_bytes {
+            junctions
+                .push(item)
+                .map_err(|_| revert("overflow when reading junctions"))?;
+        }
+
+        Ok(junctions)
+    }
+
+    fn write(writer: &mut EvmDataWriter, value: Self) {
+        let encoded: Vec<Junction> = value.iter().map(|junction| junction.clone()).collect();
+        EvmData::write(writer, encoded);
+    }
+
+    fn has_static_size() -> bool {
+        false
+    }
+}
+
+// Cannot used derive macro since it is a foreign struct.
+impl EvmData for MultiLocation {
+    fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+        let (parents, interior) = reader.read()?;
+        Ok(MultiLocation { parents, interior })
+    }
+
+    fn write(writer: &mut EvmDataWriter, value: Self) {
+        EvmData::write(writer, (value.parents, value.interior));
+    }
+
+    fn has_static_size() -> bool {
+        <(u8, Junctions)>::has_static_size()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct WeightV2 {
+    ref_time: u64,
+    proof_size: u64,
+}
+impl WeightV2 {
+    pub fn from(ref_time: u64, proof_size: u64) -> Self {
+        WeightV2 {
+            ref_time,
+            proof_size,
+        }
+    }
+    pub fn get_weight(&self) -> Weight {
+        Weight::from_parts(self.ref_time, self.proof_size)
+    }
+    pub fn is_max(&self) -> bool {
+        self.ref_time == u64::MAX
+    }
+}
+impl EvmData for WeightV2 {
+    fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+        let (ref_time, proof_size) = reader.read()?;
+        Ok(WeightV2 {
+            ref_time,
+            proof_size,
+        })
+    }
+
+    fn write(writer: &mut EvmDataWriter, value: Self) {
+        EvmData::write(writer, (value.ref_time, value.proof_size));
+    }
+
+    fn has_static_size() -> bool {
+        <(U256, U256)>::has_static_size()
+    }
+}
+#[derive(Debug)]
+pub struct EvmMultiAsset {
```

Comment:
sure.

Reply:
What is "sure"? I don't see any update on this one. You are defining everything else with no `Evm` prefix but only for MultiAsset?

Reply:
I don't see any update on this one.

Reply:
I thought you meant to make it consistent with the struct in .sol file which I did. I suppose you want me to change into `MultiAsset` but then we will clash with the `MultiAsset` in xcm and we will have to explicitly mention every where which one we want to use.
Is this what you are suggesting?

Reply:
I meant to make it without the `Evm` prefix, to be consistent with other types like `MultiLocation`.

Reply:
The reason I don't want to use `MultiAsset` is because then there will be 2 struct names `MultiAsset` which will be confusing, should I change the name to something else or should I use `MultiAsset` and do it anyways?
Kindly suggest correct approach.

Reply:
I'm fine with `EvmMultiAsset` here. But pls use `MultiAsset` in `.sol` file. It's a public interface there which is more important to keep consistent.